packages/jwt-nextjs/src/index.ts

```diff
@@ -0,0 +1,91 @@
+import { getRefreshToken, isJwtExpiredError, JwtService } from "@dotkomonline/jwt-crypto"
+import { NextRequest, NextResponse } from "next/server"
+import { DefaultJWT, getToken } from "next-auth/jwt"
+
+declare module "next-auth/jwt" {
+  interface JWT extends Record<string, unknown>, DefaultJWT {
+    accessToken?: string
+    refreshToken?: string
+  }
+}
+
+export interface MiddlewareOptions {
+  issuer: string
+  clientId: string
+  clientSecret: string
+  gatewayUrl: string
+  proxyBase?: string
+  targetBase?: string
+}
+
+export const createMiddleware = ({
+  issuer,
+  clientId,
+  clientSecret,
+  gatewayUrl,
+  proxyBase = "/api/proxy/trpc",
+  targetBase = "/api/trpc",
+}: MiddlewareOptions) => {
+  // It's important that each instatiation of the middleware has its own
+  // instance of the JwtService, as it could technically vary across different
+  // instances of the middleware.
+  const jwtService = new JwtService(issuer)
+  const getCompleteProxyUrl = (req: NextRequest): URL => {
+    const source = new URL(req.url)
+    const relativePathName = source.pathname.replace(proxyBase, targetBase)
+    const target = new URL(relativePathName, gatewayUrl)
+    for (const [key, value] of source.searchParams) {
+      target.searchParams.set(key, value)
+    }
+    return target
+  }
+
+  return async function middleware(req: NextRequest): Promise<NextResponse> {
+    const endpoint = getCompleteProxyUrl(req)
+    const token = await getToken({ req })
+    // If this is an unauthenticated request, we'll just pass it through to the
+    // trpc server without any modifications.
+    if (token === null || token?.accessToken === undefined) {
+      return NextResponse.rewrite(endpoint, {
+        request: {
+          headers: req.headers,
+        },
+      })
+    }
+
+    // Otherwise, we'll verify the token, and optionally refresh it if it has
+    // expired. Then we'll pass the request through to the trpc server.
+    try {
+      // Attempt to verify the token, and forward the request if successful.
+      await jwtService.verify(token.accessToken)
+      req.headers.set("Authorization", `Bearer ${token.accessToken}`)
+      return NextResponse.rewrite(endpoint, {
+        request: {
+          headers: req.headers,
+        },
+      })
+    } catch (error) {
+      // If the token has expired, and we have a refresh token, we'll attempt to
+      // refresh the token and forward the request with the new token.
+      if (isJwtExpiredError(error) && token.refreshToken !== undefined) {
+        const accessToken = await getRefreshToken({
+          issuer,
+          refreshToken: token.refreshToken,
+          clientId,
+          clientSecret,
+        })
+        req.headers.set("Authorization", `Bearer ${accessToken}`)
+        return NextResponse.rewrite(endpoint, {
+          request: {
+            headers: req.headers,
+          },
+        })
+      }
```

Comment:
https://auth0.com/docs/secure/tokens/refresh-tokens/refresh-token-rotation 

I think this applies even if Monoweb is not conf'd as a SPA, as it is an example in the [OAuth2 RFC Refresh Tokens Security Considerations](https://datatracker.ietf.org/doc/html/rfc6749#section-10.4) 

By requesting a new access token you also get a new refresh-token, which invalidates the refresh-token you used. This appears to not store the new request-token in the nextAuth session, so this does not work as you want, and users will get logged out in the case the code runs since the nextAuth-session will have a outdated access-token and an invalid refresh-token.

I do not think we should be dealing with refresh tokens here, does nextAuth not already handle this?

Reply:
Yeah you are probably right that this isn't the place. I will do what https://next-auth.js.org/v3/tutorials/refresh-token-rotation explains